Fix missing copy of nonlocal_vector in `TpetraWrappers::Vector` assignment operator

QY-Shi opened the pull request:
Fixed a bug in `TpetraWrappers::Vector::operator=()` where the `nonlocal_vector` (which is used to store ghost element contributions) was not copied when the destination vector was uninitialized. This could lead to missing ghost data and incorrect behavior during `compress()` operations.

The fix ensures that if the source vector contains a non-null nonlocal_vector, it is also deep-copied into the destination.

QY-Shi commented:
However, the `operator=()` function contains multiple if-else branches depending on the layout and size of the vectors. Should we consider copying the nonlocal_vector in those other branches as well (e.g., when the maps are the same or when using import/export)?


And from a design perspective, manually maintaining a separate `nonlocal_vector` feels both error-prone and cumbersome. I noticed that Trilinos provides [Tpetra::FECrsMatrix](https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1FECrsMatrix.html) and `Tpetra::FEMultiVector`—although they are not very well documented—which are designed for finite element assembly and internally handle nonlocal entries. Would it be more robust in the long term to consider using these classes instead of duplicating similar 'nonlocal' logic manually?

bangerth commented:
@kinnewig @jpthiele Do you have opinions about this?

We could perhaps also enforce via an assertion that if you copy a vector, the rhs vector must not have nonlocal elements that have not yet been communicated via `compress()`. I think that would be a sensible requirement.

kronbichler commented:
> And from a design perspective, manually maintaining a separate nonlocal_vector feels both error-prone and cumbersome. I noticed that Trilinos provides [Tpetra::FECrsMatrix](https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1FECrsMatrix.html) and Tpetra::FEMultiVector—although they are not very well documented—which are designed for finite element assembly and internally handle nonlocal entries. Would it be more robust in the long term to consider using these classes instead of duplicating similar 'nonlocal' logic manually?

I do not follow all recent developments in Tpetra closely, so I can't say about the status; however some years ago the general recommendation was to maintain the two vectors if the data is available on the application side (i.e., deal.II). The `FE`-type interfaces were at least at some point in development provided for difficult cases where no pre-computation of the sparsity and communication patterns are possible. In particular if one intends to work on the GPU, the precomputation of the data structures into these two separate matrices is crucial because one does not want to manually allocate entries on demand. In fact, in the work on our 2016 workstream paper (Turcksin et al., ACM TOMS) we found for the older Epetra stack that the `FECrsMatrix` is not scalable because of the memory allocations involved in the latter approach. It could be that the present interfaces of `Tpetra::FECrsMatrix` and `Tpetra::FEMultiVector` support the local/nonlocal matrices in finalized `FillComplete` state, where the effect would be exactly the same as what we do in our wrapper class. In that case, I would agree with you to use what is provided by Trilinos. Otherwise, I also do not see that as an overly heavy burden for us to keep those data structures.

jpthiele commented:
As for the FEXYYZ classes: 
The Epetra stack was not doing efficient storage of the nonlocal entries and therefore the non-FE classes were used.
I think there are some comments about that in the wrapper source files.
Additionally, Tpetra did not provide the FEXYZ classes for quite a while
Therefore, @kinnewig and me decided to use the non-FE classes to get a working state.

However, we have since learned that demand for a proper implementation was high 
to increase usability of Tpetra and that we should at some point rewrite our wrappers to use the FE classes.
But so far we did not have the time to do so.

QY-Shi commented:
Thank you all for the helpful replies, and for all the years of development that made these wrappers robust and usable!

For now, the current wrapper design already works well for my use case. But I definitely look forward to future updates, and I’d be happy to contribute in any way I can to support those improvements.

bangerth commented:
We got off on a different track. What do people think about copying the buffer as opposed to just requiring (via an `Assert`) that the buffer is empty upon copy?

jpthiele commented:
I can't really think of a use case for copy-assigning an uncompressed (and therefore unsynchronized) vector, 
so I feel like this might have some potential for shooting yourself in the foot.
So I would tend towards an assertion, but I am happy to be convinced by a sensible use case.

QY-Shi commented:
> Would that be a solution to your problem as well? Would you be willing to implement that instead?

No problem, I've updated the code to add an assertion as discussed.

> So I would tend towards an assertion, but I am happy to be convinced by a sensible use case.

Regarding the original use case: I initially tried to use `operator=()` to initialize an uninitialized distributed vector that had writable ghost entries (not requiring the actual values from the `nonlocal_vector`). This pattern worked with dealii::Vector and the PETSc wrapper vector, but not with Tpetra, which is why I proposed copying the nonlocal_vector as well.

However, I agree this use case may lead to subtle bugs. So I've removed the copy logic and added the assertion instead, and I'll also change my code to avoid relying on operator= for such initialization.

masterleinad commented:
/rebuild

bangerth reviewed (approved):
Yes, this looks good. Thank you!
